google/cloud/bigquery_storage_v1/writer.py

```diff
@@ -106,154 +106,44 @@ def __init__(
                 Extra headers to include when sending the streaming request.
         """
         self._client = client
-        self._closing = threading.Lock()
         self._closed = False
         self._close_callbacks = []
-        self._futures_queue = queue.Queue()
         self._metadata = metadata
+        self._thread_lock = threading.RLock()
+
+        self._stream_name = None
 
         # Make a deepcopy of the template and clear the proto3-only fields
         self._initial_request_template = _process_request_template(
             initial_request_template
         )
 
-        # Only one call to `send()` should attempt to open the RPC.
-        self._opening = threading.Lock()
-
-        self._rpc = None
-        self._stream_name = None
-
-        # The threads created in ``._open()``.
-        self._consumer = None
+        self._connection = _Connection(
+            client=client,
+            writer=self,
+            metadata=metadata,
+        )
 
     @property
     def is_active(self) -> bool:
-        """bool: True if this manager is actively streaming.
+        """bool: True if this manager is actively streaming. It is recommended
+        to call this property inside a thread lock to avoid any race conditions.
 
         Note that ``False`` does not indicate this is complete shut down,
         just that it stopped getting new messages.
         """
-        return self._consumer is not None and self._consumer.is_active
+        return self._connection is not None and self._connection.is_active
```

Comment:
Should this access to self._connection have read lock?

Reply:
Somewhat related to https://github.com/googleapis/python-bigquery-storage/pull/901#discussion_r1992066031, having the lock wouldn't prevent a race condition from happening. The user would have to call the method from inside a lock, and we have pointed it out in the docstring.